numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -426,16 +440,25 @@ UNICODE_setitem(PyObject *op, void *ov, void *vap)
     char *buffer;
 #endif
 
-    if (!PyBytes_Check(op) && !PyUnicode_Check(op) &&
-            PySequence_Check(op) && PySequence_Size(op) > 0) {
+    /* Handle case of assigning from an array scalar */
+    if (PyArray_Check(op) && PyArray_NDIM((PyArrayObject *)op) == 0) {
```

Comment:
as this is copied on longdouble, it might make sense to move this into a function with the setitem function as argument